mipengine/controller/controller.py

```diff
@@ -109,51 +210,88 @@ def get_all_available_datasets_per_data_model(self):
-        local_nodes = node_registry.get_nodes_with_any_of_datasets(
-            data_model=data_model,
-            datasets=datasets,
+        global_node_tasks_handler = _create_node_task_handler(
+            _NodeInfoDTO(
+                node_id=global_node.id,
+                queue_address=":".join([str(global_node.ip), str(global_node.port)]),
+                db_address=":".join([str(global_node.db_ip), str(global_node.db_port)]),
+                tasks_timeout=controller_config.rabbitmq.celery_tasks_timeout,
+            )
+        )
+
+        # Get only the relevant nodes from the node registry
+        local_nodes_info = _get_nodes_info_by_dataset(
+            data_model=data_model, datasets=datasets
         )
+        local_nodes_tasks_handlers = [
+            _create_node_task_handler(task_handler) for task_handler in local_nodes_info
+        ]
 
-        queue_address = ":".join([str(global_node.ip), str(global_node.port)])
-        db_address = ":".join([str(global_node.db_ip), str(global_node.db_port)])
-        tasks_timeout = controller_config.rabbitmq.celery_tasks_timeout
-        # Instantiate the INodeTasksHandler for the Global Node
-        global_node_tasks_handler = NodeTasksHandlerCelery(
-            node_id=global_node.id,
-            node_queue_addr=queue_address,
-            node_db_addr=db_address,
-            tasks_timeout=tasks_timeout,
+        return NodesTasksHandlersDTO(
+            global_node_tasks_handler=global_node_tasks_handler,
+            local_nodes_tasks_handlers=local_nodes_tasks_handlers,
         )
 
-        local_nodes_tasks_handlers = []
-        for local_node in local_nodes:
-            # Instantiate the INodeTasksHandlers for the Local Nodes
-            queue_address = ":".join([str(local_node.ip), str(local_node.port)])
-            db_address = ":".join([str(local_node.db_ip), str(local_node.db_port)])
-            tasks_timeout = controller_config.rabbitmq.celery_tasks_timeout
 
-            node_tasks_handler = NodeTasksHandlerCelery(
-                node_id=local_node.id,
-                node_queue_addr=queue_address,
-                node_db_addr=db_address,
-                tasks_timeout=tasks_timeout,
+class _NodeInfoDTO(BaseModel):
+    node_id: str
+    queue_address: str
+    db_address: str
+    tasks_timeout: int
+
+    class Config:
+        allow_mutation = False
+
+
+def _get_node_info_by_id(node_id: str) -> _NodeInfoDTO:
+    global_nodes = node_registry.get_all_global_nodes()
+    local_nodes = node_registry.get_all_local_nodes()
+
+    for node in global_nodes + local_nodes:
+        if node.id == node_id:
+            return _NodeInfoDTO(
+                node_id=node.id,
+                queue_address=":".join([str(node.ip), str(node.port)]),
+                db_address=":".join([str(node.db_ip), str(node.db_port)]),
+                tasks_timeout=controller_config.rabbitmq.celery_tasks_timeout,
```

Comment:
If the node with `node_id` is unreachable or not in the node registry, then the  `_get_node_info_by_id(..)` function will return a `NoneType` which will then cause the subsequent call to `_create_node_task_handler` to raise an exception which will be caught in the try block line ~123. It will log that the cleanup for this node_id failed and continue. Then, after `CLEANUP_INTERVAL` seconds it will retry to cleanup this node and if it fails again it will log again the failure and continue. 
```
for node_id in node_ids:
    try:
        node_info = self._get_node_info_by_id(node_id)
        task_handler = _create_node_task_handler(node_info)
        task_handler.clean_up(
                request_id=CONTROLLER_CLEANUP_REQUEST_ID,
                context_id=context_id,
            )

        self._controller_logger.debug(
            f"clean_up task succeeded for {node_id=} for {context_id=}"
        )
        cleaned_up_nodes[context_id].append(node_id)
    except Exception as exc:
        self._controller_logger.debug(
                f"clean_up task FAILED for {node_id=} "
                f"for {context_id=}. Will retry in a while... fail "
                f"reason: {type(exc)}:{exc}"
        )
```

So it will continue to try to cleanup this node until it succeeds. I wanted to implement another mechanism there so the controller will stop trying to cleanup a specific node if it does not come online again after a prolonged amount of time, ex. some hours, but we have not discussed this scenario. So for now it will just try to cleanup the node forever and until then it will log(as debug) that it fails.

Concerning the testing of this, this is what the `test_cleanup_after_algorithm_execution.py::test_cleanup_node_down_algorithm_execution` tests. It starts an algorithm execution and before it completes it removes the rabbitmq container of the "localnodetmp" node, since the rabbitmq container of the node is down the node will not be present in the next node registry "scan", so when the cleanup process tries to cleanup all the nodes that participated in the failed algorithm execution it will fail to clean the "localnodetmp" node,but it will continue to try to clean it again every CLEANUP_INTERVAL seconds. Then, the rabbitmq container of "localnodetmp" is brought back up, the node registry picks the "localnodetmp" up and the cleanup finally succeeds.

Reply:
Makes sense, haven't seen that it silently handles the exception.

About the test, it's not certain that the node registry won't have the node still, but I guess it's not a problem with the approach currently implemented. Whatever fails the cleanup will retry, later.